### ISSUE TITLE
Push Prompt Immediately Requested

### ISSUE THREAD

**Issue by AdiAyyakad** [urbanairship/ios-library#164]
Hey so I was trying to implement your SDK using Cocoapods (Cocoapods 1.4.0 and UrbanAirship 8.6.3), and on launch, directly after running `UAirship.takeOff(UAConfig.default())` I see:
```
(lldb) po UAirship.push().userPushNotificationsEnabled
true
```
Despite the fact that your docs say that this defaults to `NO`. https://docs.urbanairship.com/reference/libraries/ios/latest/Classes/UAPush.html#/c:objc(cs)UAPush(py)userPushNotificationsEnabled

My `AirshipConfig.plist`:
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>developmentAppKey</key>
	<string>secret-app-key</string>
	<key>developmentAppSecret</key>
	<string>secret-dev-app-key</string>
	<key>productionAppKey</key>
	<string>secret-prod-key-bBhQgA</string>
	<key>productionAppSecret</key>
	<string>secret-prod-app-key</string>
	<key>useWKWebView</key>
	<true/>
    <key>automaticSetupEnabled</key>
    <false/>
</dict>
</plist>
```

Is there anything I'm missing? It's causing the system remote notifications prompt to come up on launch.

**Comment by rlepinski** [comment]
@AdiAyyakad Its suppose to be `NO` by default. Could you provide verbose logs?

**Comment by AdiAyyakad** [comment]
@rlepinski Thanks for the prompt response! I'm working on getting the logs, but I seem to have narrowed it down to this:
```
if ([self.dataStore objectForKey:UAPushEnabledKey]) { // This is true. But why?
            BOOL previousValue = [self.dataStore boolForKey:UAPushEnabledKey];
            UA_LDEBUG(@"Migrating userPushNotificationEnabled to %@ from previous pushEnabledValue.", previousValue ? @"YES" : @"NO");
            [self.dataStore setBool:previousValue forKey:UAUserPushNotificationsEnabledKey];
            [self.dataStore removeObjectForKey:UAPushEnabledKey];
        } else {
            // If >= iOS 10
            if ([[NSProcessInfo processInfo] isOperatingSystemAtLeastVersion:(NSOperatingSystemVersion){10, 0, 0}]) {
                [[UNUserNotificationCenter currentNotificationCenter] getNotificationSettingsWithCompletionHandler:^(UNNotificationSettings * _Nonnull settings) {
                    if (settings.authorizationStatus == UNAuthorizationStatusAuthorized) {
                        UA_LDEBUG(@"Migrating userPushNotificationEnabled to YES because application was authorized for notifications");
                        [self.dataStore setBool:YES forKey:UAUserPushNotificationsEnabledKey];
                    }
                }];
```

**Comment by AdiAyyakad** [comment]
The dataStore seems to be setting `UAPushEnabledKey` to `true` by default, and so this goes through and prompts on launch

**Comment by AdiAyyakad** [comment]
@rlepinski 
```
2018-01-24 17:04:31.882615-0500 LAT[367:30527] [DYMTLInitPlatform] platform initialization successful
\2018-01-24 17:04:32.533036-0500 LAT[367:30275] [I] +[UAirship executeUnsafeTakeOff:] [Line 185] UAirship Take Off! Lib Version: 8.6.3 App Key: ~redacted~ Production: NO.
2018-01-24 17:04:32.533462-0500 LAT[367:30275] [T] +[UAirship executeUnsafeTakeOff:] [Line 194] Registering UAURLProtocol.
2018-01-24 17:04:32.541402-0500 LAT[367:30275] [T] +[UAKeychainUtils getDeviceID] [Line 190] Retrieved Device ID info from keychain.
2018-01-24 17:04:32.541427-0500 LAT[367:30275] [T] +[UAKeychainUtils getDeviceID] [Line 217] Loaded Device ID: ~redacted~
2018-01-24 17:04:32.552404-0500 LAT[367:30275] [D] -[UANamedUser update] [Line 65] New or re-install, skipping named user update.
2018-01-24 17:04:32.576287-0500 LAT[367:30511] [I] -[UAInboxMessageList refreshInboxWithCompletionHandler:]_block_invoke [Line 295] Inbox messages updated.
2018-01-24 17:04:32.576356-0500 LAT[367:30511] [T] -[UAInboxMessageList refreshInboxWithCompletionHandler:]_block_invoke [Line 297] Loaded saved messages: (
).
2018-01-24 17:04:32.610850-0500 LAT[367:30275] [D] -[UAAnalytics addEvent:] [Line 176] Adding app_init event ~redacted~.
2018-01-24 17:04:32.610918-0500 LAT[367:30275] [D] -[UAEventManager scheduleUploadWithDelay:] [Line 269] Enqueuing attempt to schedule event upload with delay on main queue.
2018-01-24 17:04:32.611123-0500 LAT[367:30275] [T] -[UAAnalytics addEvent:] [Line 178] Event added: UAEvent ID: ~redacted~ type: app_init time: 1516831472.608243 data: {
    ~redacted~
}.
2018-01-24 17:04:32.611564-0500 LAT[367:30540] [T] -[UAEventStore storeEventWithID:eventType:eventTime:eventBody:sessionID:] [Line 273] Event saved: ~redacted~
2018-01-24 17:04:32.810619-0500 LAT[367:30275] [D] -[UAEventManager scheduleUploadWithDelay:]_block_invoke [Line 272] Attempting to schedule event upload with delay: 14.944039 seconds.
2018-01-24 17:04:32.810648-0500 LAT[367:30275] [D] -[UAEventManager scheduleUploadWithDelay:]_block_invoke [Line 286] Scheduling upload.
2018-01-24 17:04:32.811143-0500 LAT[367:30275] [D] -[UAPush updateRegistration] [Line 778] APNS registration is out of date, updating.
2018-01-24 17:04:32.812053-0500 LAT[367:30275] [T] -[UALocation locationManager:didChangeAuthorizationStatus:] [Line 212] Location authorization changed: 0
2018-01-24 17:04:32.815145-0500 LAT[367:30275] [D] -[UAPush updateRegistration] [Line 778] APNS registration is out of date, updating.
2018-01-24 17:04:32.817855-0500 LAT[367:30275] [I] +[UAAppIntegration application:didRegisterForRemoteNotificationsWithDeviceToken:] [Line 43] Application registered device token: ~redacted~
2018-01-24 17:04:32.817933-0500 LAT[367:30275] [D] -[UAAnalytics addEvent:] [Line 176] Adding device_registration event ~redacted~.
2018-01-24 17:04:32.817984-0500 LAT[367:30275] [D] -[UAEventManager scheduleUploadWithDelay:] [Line 269] Enqueuing attempt to schedule event upload with delay on main queue.
2018-01-24 17:04:32.818031-0500 LAT[367:30569] [T] -[UAEventStore storeEventWithID:eventType:eventTime:eventBody:sessionID:] [Line 273] Event saved: ~redacted~
2018-01-24 17:04:32.818057-0500 LAT[367:30275] [T] -[UAAnalytics addEvent:] [Line 178] Event added: UAEvent ID: ~redacted~ type: device_registration time: 1516831472.817888 data: {
    ~redacted~
}.
2018-01-24 17:04:32.818184-0500 LAT[367:30275] Device token: ~redacted~
2018-01-24 17:04:32.829742-0500 LAT[367:30275] [D] -[UAChannelRegistrar createChannelWithPayload:] [Line 135] Creating channel.
2018-01-24 17:04:32.829913-0500 LAT[367:30275] [T] -[UAChannelAPIClient createChannelWithPayload:onSuccess:onFailure:] [Line 27] Creating channel with: {
    ~redacted~
}.
2018-01-24 17:04:32.831047-0500 LAT[367:30275] [D] -[UAEventManager scheduleUploadWithDelay:]_block_invoke [Line 272] Attempting to schedule event upload with delay: 14.736973 seconds.
2018-01-24 17:04:32.831071-0500 LAT[367:30275] [D] -[UAEventManager scheduleUploadWithDelay:]_block_invoke [Line 277] Upload already scheduled for an earlier time.
2018-01-24 17:04:32.894718-0500 LAT[367:30275] [T] -[UAAnalytics didBecomeActive] [Line 145] Application did become active.
2018-01-24 17:04:32.894935-0500 LAT[367:30275] [D] -[UAInboxMessageList retrieveMessageListWithSuccessBlock:withFailureBlock:] [Line 97] Retrieving message list.
2018-01-24 17:04:32.894986-0500 LAT[367:30275] [T] -[UAInboxAPIClient retrieveMessageListOnSuccess:onFailure:]_block_invoke [Line 61] Request to retrieve message list: https://device-api.urbanairship.com/api/user/~redacted~/messages/
2018-01-24 17:04:33.146387-0500 LAT[367:30569] [T] -[UAChannelAPIClient createChannelWithPayload:onSuccess:onFailure:]_block_invoke_2 [Line 67] Channel creation succeeded with status: 200 jsonResponse: {
    ~redacted~
}
2018-01-24 17:04:33.146442-0500 LAT[367:30569] [D] -[UAChannelRegistrar createChannelWithPayload:]_block_invoke [Line 147] Channel ~redacted~ created successfully. Channel location: https://device-api.urbanairship.com/api/channels/~redacted~.
2018-01-24 17:04:33.146650-0500 LAT[367:30275] Channel ID: ~redacted~
2018-01-24 17:04:33.146837-0500 LAT[367:30275] Created channel with ID: ~redacted~
2018-01-24 17:04:33.147067-0500 LAT[367:30275] [T] -[UAUser updateUser] [Line 207] Updating user
2018-01-24 17:04:33.147163-0500 LAT[367:30275] [D] -[UAUserAPIClient requestToUpdateUser:payload:]_block_invoke [Line 163] Request to update user with body: <~redacted~>
2018-01-24 17:04:33.147289-0500 LAT[367:30275] [D] -[UANamedUser update] [Line 65] New or re-install, skipping named user update.
2018-01-24 17:04:33.147330-0500 LAT[367:30275] [D] -[UAEventManager scheduleUploadWithDelay:] [Line 269] Enqueuing attempt to schedule event upload with delay on main queue.
2018-01-24 17:04:33.147354-0500 LAT[367:30275] [I] -[UAPush registrationSucceededWithPayload:] [Line 921] Channel registration updated successfully.
2018-01-24 17:04:33.151351-0500 LAT[367:30275] [D] -[UAEventManager scheduleUploadWithDelay:]_block_invoke [Line 272] Attempting to schedule event upload with delay: 14.747375 seconds.
2018-01-24 17:04:33.151480-0500 LAT[367:30275] [D] -[UAEventManager scheduleUploadWithDelay:]_block_invoke [Line 286] Scheduling upload.
2018-01-24 17:04:33.170304-0500 LAT[367:30569] [T] -[UAInboxAPIClient retrieveMessageListOnSuccess:onFailure:]_block_invoke_2 [Line 109] Retrieved message list with status: 200 jsonResponse: {
    badge = 0;
    messages =     (
    );
}
2018-01-24 17:04:33.170358-0500 LAT[367:30569] [D] -[UAInboxAPIClient retrieveMessageListOnSuccess:onFailure:]_block_invoke_2 [Line 111] Setting Last-Modified time to '(null)' for user ~redacted~'s message list.
2018-01-24 17:04:33.170647-0500 LAT[367:30569] [D] -[UAInboxMessageList retrieveMessageListWithSuccessBlock:withFailureBlock:]_block_invoke [Line 135] Refreshing message list.
2018-01-24 17:04:33.178543-0500 LAT[367:30537] [I] -[UAInboxMessageList refreshInboxWithCompletionHandler:]_block_invoke [Line 295] Inbox messages updated.
2018-01-24 17:04:33.178600-0500 LAT[367:30537] [T] -[UAInboxMessageList refreshInboxWithCompletionHandler:]_block_invoke [Line 297] Loaded saved messages: (
).
2018-01-24 17:04:33.740723-0500 LAT[367:30537] [T] -[UAUserAPIClient updateUser:channelID:onSuccess:onFailure:]_block_invoke_2 [Line 115] Successfully updated user: <UAUser: 0x1d028c260>
2018-01-24 17:04:33.740762-0500 LAT[367:30537] [I] -[UAUser updateUser]_block_invoke [Line 211] Updated user ~redacted~ successfully.
2018-01-24 17:04:47.905274-0500 LAT[367:30511] [D] -[UAEventManager enqueueUploadOperationWithDelay:]_block_invoke [Line 302] Preparing events for upload
2018-01-24 17:04:47.915845-0500 LAT[367:30569] [T] -[UAEventManager enqueueUploadOperationWithDelay:]_block_invoke_2 [Line 354] Uploading events.
2018-01-24 17:04:47.923730-0500 LAT[367:30275] [T] -[UAEventAPIClient uploadEvents:completionHandler:] [Line 31] Sending analytics events with IDs: (
~redacted~
)
2018-01-24 17:04:47.923823-0500 LAT[367:30275] [T] -[UAEventAPIClient uploadEvents:completionHandler:] [Line 32] Sending to server: https://combine.urbanairship.com
2018-01-24 17:04:47.924168-0500 LAT[367:30275] [T] -[UAEventAPIClient uploadEvents:completionHandler:] [Line 33] Sending analytics headers:     {
~redacted~
    }
2018-01-24 17:04:47.925190-0500 LAT[367:30275] [T] -[UAEventAPIClient uploadEvents:completionHandler:] [Line 34] Sending analytics body: [
~redacted~
]
2018-01-24 17:04:48.205370-0500 LAT[367:30652] [D] -[UAEventManager enqueueUploadOperationWithDelay:]_block_invoke_4 [Line 372] Analytic upload success
2018-01-24 17:04:48.206644-0500 LAT[367:30652] [T] -[UAEventManager enqueueUploadOperationWithDelay:]_block_invoke_4 [Line 373] Response: <NSHTTPURLResponse: 0x1cc03e800> { URL: https://combine.urbanairship.com/warp9/ } { Status Code: 200, Headers {
~redacted~
} }
```

Here are some logs that are printed out. But this

**Comment by rlepinski** [comment]
Are you coming from an older SDK version or is this a new integration?

**Comment by AdiAyyakad** [comment]
New integration. This is off a clean build - after deleting the app and then building on device.

**Comment by rlepinski** [comment]
@AdiAyyakad I suspect its the code you pointed out, but not seeing the log message `Migrating userPushNotificationEnabled`, could you provide logs for the session that it migrates the value. Not really sure whats going on here but I will try to reproduce in our sample application. 

Looks like the key is "UAPushEnabled", are you by any chance setting that somewhere? Did you come from a different push lib that I can go check out?

**Comment by AdiAyyakad** [comment]
@rlepinski That's not being set anywhere from my knowledge. All we did was just pod install the lib and added the AirshipConfig file I pasted above. I tried searching for that key and everything related to it and our code doesn't set it. And no we're not coming from any other push library - just `pod 'UrbanAirship-iOS-SDK'`

**Comment by rlepinski** [comment]
@AdiAyyakad I am unable to reproduce it with my sample app, but I might have a work around for you until we can figure this issue out. Try to set `@"UAPushEnabledSettingsMigrated"` to `YES` in user defaults before you call takeOff.

**Comment by AdiAyyakad** [comment]
I just tried running the app cleanly without `UAirship.takeOff(UAConfig.default())` and nothing was prompted. I added it back in and prompt.

I put a breakpoint on `[UNUserNotificationCenter requestAuthorization]` and it was hit in `UAAPNSRegistration.m` on launch.
```
-(void)updateRegistrationWithOptions:(UANotificationOptions)options
                          categories:(NSSet<UANotificationCategory *> *)categories {

#if !TARGET_OS_TV   // UNNotificationCategory not supported on tvOS
    NSMutableSet *normalizedCategories;

    if (categories) {
        normalizedCategories = [NSMutableSet set];

        // Normalize our abstract categories to iOS-appropriate type
        for (UANotificationCategory *category in categories) {

            id normalizedCategory = [category asUNNotificationCategory];

            // iOS 10 beta this could return nil
            if (normalizedCategory) {
                [normalizedCategories addObject:normalizedCategory];
            }
        }
    }

    [[UNUserNotificationCenter currentNotificationCenter] setNotificationCategories:[NSSet setWithSet:normalizedCategories]];
#endif

    UNAuthorizationOptions normalizedOptions = (UNAuthorizationOptionAlert | UNAuthorizationOptionBadge | UNAuthorizationOptionSound | UNAuthorizationOptionCarPlay);
    normalizedOptions &= options;

    [[UNUserNotificationCenter currentNotificationCenter] requestAuthorizationWithOptions:normalizedOptions
                                                                        completionHandler:^(BOOL granted, NSError * _Nullable error) {
                                                                            [self getCurrentAuthorizationOptionsWithCompletionHandler:^(UANotificationOptions authorizedOptions) {
                                                                                [self.registrationDelegate notificationRegistrationFinishedWithOptions:authorizedOptions];
                                                                            }];
                                                                        }];
}
```

**Comment by rlepinski** [comment]
@AdiAyyakad Any chance you could provide a sample app that has the issue? Have you tried a different simulator/device?

**Comment by AdiAyyakad** [comment]
I've tried two devices (iPhone X and iPad both running 11.2.5), and on simulators, I'm not getting the prompt, but I immediately get a notification to `didfailtoregisterforremotenotificationswitherror` so it's getting asked. I tried to create a sample project that illustrates the issue but I can't seem to reproduce that happening anywhere but our app unfortunately.

**Comment by AdiAyyakad** [comment]
@rlepinski Oh I narrowed it down! So on launch, we update device token if `UIApplication.shared.isRegisteredForRemoteNotifications` is `true`. It seems that after `[UAirship executeUnsafeTakeOff:[config copy]]` is run, then that `UIApplication.shared.isRegisteredForRemoteNotifications` (which is `false` before this is run), is set to `true`!

**Comment by rlepinski** [comment]
@AdiAyyakad Since iOS 9, Apple allows apps to request user notifications separately from registering with APNS. This allows apps to have background/content pushes without notifications. This call should not be causing a prompt.

**Comment by AdiAyyakad** [comment]
@rlepinski Understood! The prompt is definitely happening on our end - but it's only because we've been assuming that we'll only register for remote notifications when a user agrees to our pre-prompt. Is it possible to disable this on UA end?

**Comment by rlepinski** [comment]
Trying to get more info, so when you get a successful callback for APNs registration your application tries to register for notifications by calling the `UNUserNotificationCenter` directly?

**Comment by AdiAyyakad** [comment]
No so on `didFinishLaunchingWithOptions`, after we call `takeOff`, we then try to update the device token. This is because we should be updating the device token on launch and just that's how our app was written pre-UA integration. so we just do this:
```
guard UIApplication.shared.isRegisteredForRemoteNotifications else {
    return // Is not currently registered. Do not update device token.
}

// Registration flow
```

**Comment by AdiAyyakad** [comment]
After consulting with my team, we realized we'll just change that registration flow to just update the remote notification and hold off on prompting the user till later. Sorry for the confusion and thanks for helping me debug the solution!

**Comment by rlepinski** [comment]
@AdiAyyakad Glad you figured it out.

We try to provide an abstraction from apples APNS/notification registration flows to be backwards compatible with older versions and to make sure our registration flow is doing well. To prompt a user to for notifications, its preferred to just enable notification with `UAirship.push().userPushNotificationsEnabled = true` and to set any custom setting through our UAPush instance so we are not conflicting with the apps calls to underlying APIs. You should check out the [registration delegate](https://docs.urbanairship.com/reference/libraries/ios/latest/Protocols/UARegistrationDelegate.html) and [push delegate.](https://docs.urbanairship.com/reference/libraries/ios/latest/Protocols/UAPushNotificationDelegate.html) both can be set on the UAPush instance. Hopefully it simplifies your app instead of complicating it.

**Comment by AdiAyyakad** [comment]
@rlepinski I'll definitely look into it I think it just might. Thanks again!

**Comment by cerupcat** [comment]
I'm running into this same issue that calling ` [UAirship takeOff]` automatically prompts and removing it doesn't prompt. I couldn't follow the above conversation completely to understand what I should be changing on our end to fix this.

**Comment by AdiAyyakad** [comment]
@cerupcat 
The issue was on our end for us. We basically had a `didFinishLaunchingWithOptions` that had a flow like this:
```
UAirship.takeOff()
guard UIApplication.shared.isRegisteredForRemoteNotifications else {
    return
}

// This registers users for both Remote and UI notifications
registerUsersForNotifications()
```

So on running `UAirship.takeOff()`, it registers your device for silent notifications automatically (can't find a way to turn it off). So after that, when we checked against being `registeredForRemoteNotifications`, it came back `true` causing us to jump into our re-registration code that was also erroneously requesting authorization for user-displayed push notifications with `UNNotificationCenter.current().requestAuthorization`. So once we removed that code, we were all set.
